dkulib/parallelizer/parallelizer.py

```diff
@@ -0,0 +1,318 @@
+        return output
+
+    def _post_process_results(
+        self, df: pd.DataFrame, results: List[Dict]
+    ) -> pd.DataFrame:
+        """Combines results from the function with the input dataframe"""
+        results = flatten(results)
+        output_schema = {
+            **{column_name: str for column_name in self._output_column_names},
+            **dict(df.dtypes),
+        }
+        output_df = (
+            pd.DataFrame.from_records(results)
+            .reindex(columns=list(df.columns) + list(self._output_column_names))
+            .astype(output_schema)
+        )
+        if not self.verbose:
+            output_df.drop(
+                labels=self._output_column_names.error_raw, axis=1, inplace=True
+            )
+        if self.error_handling == ErrorHandling.FAIL:
+            error_columns = [
+                self._output_column_names.error_message,
+                self._output_column_names.error_type,
+                self._output_column_names.error_raw,
+            ]
+            output_df.drop(labels=error_columns, axis=1, inplace=True, errors="ignore")
+        num_error = sum(output_df[self._output_column_names.response] == "")
+        num_success = len(df.index) - num_error
+        logging.info(
+            f"Applying function {self.function.__name__} in parallel to {len(df.index)} row(s): "
+            + f"{num_success} row(s) succeeded, {num_error} failed."
+        )
+        return output_df
+
+    def run(self, df: pd.DataFrame, **function_kwargs,) -> pd.DataFrame:
+        """Applies a function to a pandas.DataFrame with parallelization, error logging and progress tracking.
+
+        The DataFrame is iterated on and fed to the function as dictionaries, row-by-row or by batches of rows.
+        This process is accelerated by the use of concurrent threads and is tracked with a progress bar.
+        Errors are catched if they match the `self.exceptions_to_catch` attribute and automatically logged.
+        Once the whole DataFrame has been iterated on, results and errors are added as additional columns.
+
+        Args:
+            df: Input dataframe on which the function will be applied
+            **function_kwargs: Arbitrary keyword arguments passed to the `function`
+
+        Returns:
+            Input dataframe with additional columns:
+            - response from the `function`
+            - error message if any
+            - error type if any
+
+        """
+        # First, we create a generator expression to yield each row of the input dataframe.
+        # Each row will be represented as a dictionary like {"column_name_1": "foo", "column_name_2": 42}
+        df_row_generator = (
+            index_series_pair[1].to_dict() for index_series_pair in df.iterrows()
+        )
+        # We then chunk the generator into lists of rows of len batch_size
+        df_row_batch_generator = chunked(df_row_generator, self.batch_size)
```

Comment:
The variable `df_row_batch_generator` is not useful I would inline it

Reply:
2a197b8